src/clib/pio_darray.c

```diff
@@ -665,64 +670,90 @@ static int register_decomp(file_desc_t *file, int ioid)
-            }
-            temp_mapbuf[0] = iodesc->map[0];
-            temp_mapbuf[1] = 0;
-            mapbuf = (char*)temp_mapbuf;
+            mapbuf = (long*)malloc(sizeof(long)*maplen);
+            ((long*)mapbuf)[0] = iodesc->map[0];
+            ((long*)mapbuf)[1] = 0;
         }
 
-        int64_t vid = adios_define_var(file->adios_group, name, "", type, ldim, "", "");
-        adios_write_byid(file->adios_fh, vid, mapbuf);
+        av_count[0] = (size_t)maplen;
+        adios2_variable *variableH = adios2_inquire_variable(file->ioH, name);
+        if (variableH == NULL)
+        {
+            variableH = adios2_define_variable(file->ioH, name, type,
+                                               1, NULL, NULL, av_count,
+                                               adios2_constant_dims_true);
+        }
+        adios2_put(file->engineH, variableH, mapbuf, adios2_mode_sync);
 
         if (mapbuf != NULL)
             free(mapbuf);
     }
+    else /* Handle the case where maplen is less than 1 */
+    {
+        long mapbuf[2];
+        mapbuf[0] = 0;
+        mapbuf[1] = 0;
+        av_count[0] = (size_t)2;
+
+        adios2_variable *variableH = adios2_inquire_variable(file->ioH, name);
+        if (variableH == NULL)
+        {
+            variableH = adios2_define_variable(file->ioH, name, type,
+                                               1, NULL, NULL, av_count,
+                                               adios2_constant_dims_true);
+        }
+        adios2_put(file->engineH, variableH, mapbuf, adios2_mode_sync);
+    }
 
     /* ADIOS: assume all procs are also IO tasks */
     if (file->adios_iomaster == MPI_ROOT)
     {
-        adios_define_attribute_byvalue(file->adios_group, "piotype", name, adios_integer, 1, &iodesc->piotype);
-        adios_define_attribute_byvalue(file->adios_group, "ndims", name, adios_integer, 1, &iodesc->ndims);
-        adios_define_attribute_byvalue(file->adios_group, "dimlen", name, adios_integer, iodesc->ndims, iodesc->dimlen);
+        char att_name[PIO_MAX_NAME];
+
+        sprintf(att_name, "%s/piotype", name);
+        if (adios2_inquire_attribute(file->ioH, att_name) == NULL)
+            adios2_define_attribute(file->ioH, att_name, adios2_type_int32_t, &iodesc->piotype);
+
+        sprintf(att_name, "%s/ndims", name);
+        if (adios2_inquire_attribute(file->ioH, att_name) == NULL)
+            adios2_define_attribute(file->ioH, att_name, adios2_type_int32_t, &iodesc->ndims);
+
+        sprintf(att_name, "%s/dimlen", name);
+        if (adios2_inquire_attribute(file->ioH, att_name) == NULL)
+            adios2_define_attribute_array(file->ioH, att_name, adios2_type_int32_t, iodesc->dimlen, iodesc->ndims);
```

Comment:
Try using snprintf instead of sprintf and handle the return values from the adios2 APIs